### PR TITLE
:mortar_board: Topic22 Example general tree algorithms 

### PR DESCRIPTION
**Pull request by jameshughes89**
### Related Issues or PRs
Finishing up section from #314 

### What
Add a few example tree algorithms. 

### Why
They're good to know, but it will also allow for a comparison to the `BinarySearchTree` implementation.

### CODE REVIEW

File: src/site/topic22.rst [hunk 1]
```diff
@@ -263,6 +263,106 @@ Binary Tree Node
...
+
+    private int size(Node<T> current) {
+        if (current == null) {
+            return 0;
+        } else {
+            return 1 + size((current.getLeft())) + size(current.getRight());
+        }
+    }
+
+* The idea here is, if the current node exists, the the size of the (sub)tree will be 1 + the size of the left subtree + the size of the right subtree
```
Comment (twentylemon) [inline src/site/topic22.rst:287]: This probably makes more sense before the implementation, ie present the idea for the solution, then the code.

Similar for other operations.

File: src/site/topic22.rst [hunk 2]
```diff
@@ -263,6 +263,106 @@ Binary Tree Node
...
+        if (current == null) {
+            return false;
+        } else if (current.getData().equals(element)) {
+            return true;
+        } else {
+            boolean inSubtree = contains(current.getLeft(), element);
+            if(!inSubtree) {
+                inSubtree = contains(current.getRight(), element);
+            }
+            return inSubtree;
```
Comment (twentylemon) [inline src/site/topic22.rst:314]: I don't like the mutable value here, pretty weird implementation to read. More simply, `return contains(current.getLeft(), needle) || contains(current.getRight(), needle);`

File: src/site/topic22.rst [hunk 3]
```diff
@@ -263,6 +263,106 @@ Binary Tree Node
...
+    * :math:`O(n)`, where :math:`n` is the number of nodes in the tree
+
+**Contains**
+
+* If we have an arbitrary binary tree and we want to search it for a specific element
+
+.. code-block:: java
+    :linenos:
+
+    public boolean contains(T element) {
```
Comment (twentylemon) [inline src/site/topic22.rst:300]: `element` -> `needle` ([as mentioned in previous topics](https://github.com/jameshughes89/cs102/blame/main/src/site/topic20.rst#L18))

Comment (jameshughes89) [reply src/site/topic22.rst:300]: changed in the code, but will keep element in cases like: 

> If we have an arbitrary binary tree and we want to search it for a specific element

Comment (twentylemon) [reply src/site/topic22.rst]: I don't see why. An array is as arbitrary a collection as a tree. Search is search.

Comment (twentylemon) [reply src/site/topic22.rst]: Ah, I misunderstood. You meant in the text, you'd use the word element. That's fine, makes sense.

**Review by twentylemon — APPROVED** [review]

File: src/site/topic22.rst (hunk 3 above)
Comment (jameshughes89) [reply src/site/topic22.rst]: Brevity is perhaps not the best way to communicate information.

Comment (twentylemon) [reply src/site/topic22.rst]: 🤔

Comment (jameshughes89) [reply src/site/topic22.rst]: :thinking: :thinking: :thinking:

**Pull request merged**